### PR TITLE
Refactoring of pkcs11_store.c module

### PR DESCRIPTION
**Pull request by danielinux**
-  [x] add unit tests for `pkcs11_store.c`
-  [x] defects detected by unit tests are fixed
-  [x] refactoring with a new powerfail-safe design

This new design uses two sectors at the beginning of the keyvault partition.
The first sector contains the allocation tables for the nodes in flash, including
a bitmap to keep track of the free slots.
The second sector is a pre-commit 'backup' sector, used to update the content of the
vault in two steps. This guarantees powerfail protection.

All hal_flash_erase / hal_flash_write operations are now on entire sectors.
Sector content is cached to memory, then written to the backup sector, then
committed to the actual destination sector.

### CODE REVIEW

**Comment by danielinux** [comment]
Awaiting feedback (zd18366)

File: src/pkcs11_store.c [hunk 1]
```diff
@@ -97,36 +101,39 @@ int wolfPKCS11_Store_Open(int type, CK_ULONG id1, CK_ULONG id2, int read,
...
-    if ((!found) && read) {
+    if ((found == -1) && read) {
         *store = NULL;
         return NOT_AVAILABLE_E;
-    } else if (found) {
-        *store = vault_descriptors[found];
-        obj = vault_descriptors[found];
-        memcpy(&obj->hdr, vault_base + found * KEYVAULT_OBJ_SIZE, sizeof(struct obj_hdr));
+    }
+    obj = XMALLOC(sizeof(struct store_object), NULL, DYNAMIC_TYPE_TMP_BUFFER);
```
Comment (dgarske) [inline src/pkcs11_store.c:120]: Consider cast `(struct store_object*)`.

File: src/pkcs11_store.c [hunk 2]
```diff
@@ -97,36 +101,39 @@ int wolfPKCS11_Store_Open(int type, CK_ULONG id1, CK_ULONG id2, int read,
...
+    if (!obj) {
```
Comment (dgarske) [inline src/pkcs11_store.c:121]: Prefer `obj == NULL`... thus not assuming NULL == 0.

File: src/pkcs11_store.c (hunk 1 above)
Comment (dgarske) [inline src/pkcs11_store.c:120]: Consider version that doesn't require heap..

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/pkcs11_store.c (hunk 1 above)
Comment (danielinux) [reply src/pkcs11_store.c]: XMALLOC is gone now

File: src/pkcs11_store.c (hunk 2 above)
Comment (danielinux) [reply src/pkcs11_store.c]: ditto

File: src/pkcs11_store.c (hunk 1 above)
Comment (danielinux) [reply src/pkcs11_store.c]: I have removed the dynamic allocation in the last refactoring, however it seems that wolfPKCS11 still requires a heap, so I re-introduced the _sbrk in the designated pool.

Comment (danielinux) [reply src/pkcs11_store.c]: Ideally, since wolfPKCS11 uses its own area, this heap can be in a bank that is securely erased on powerdown

**Review by dgarske — DISMISSED** [review]

**Comment by danielinux** [comment]
Ready to review

**Review by dgarske — DISMISSED** [review]

**Pull request merged**